hal/src/thumbv6m/sercom/v1/i2c.rs

```diff
@@ -85,38 +91,53 @@ where
         // safe because we're exclusively owning SERCOM
         pm.apbcmask.modify(|_, w| w.$powermask().set_bit());
 
+        let mut s = Self { sda, scl, sercom };
+
         unsafe {
             // reset the sercom instance
-            sercom.i2cm().ctrla.modify(|_, w| w.swrst().set_bit());
+            s.sercom.i2cm().ctrla.modify(|_, w| w.swrst().set_bit());
             // wait for reset to complete
-            while sercom.i2cm().syncbusy.read().swrst().bit_is_set()
-                || sercom.i2cm().ctrla.read().swrst().bit_is_set()
+            while !s.is_register_synced(i2c_reg_sync::SwRst)
+                || s.sercom.i2cm().ctrla.read().swrst().bit_is_set()
             {}
 
             // Put the hardware into i2c master mode
-            sercom.i2cm().ctrla.modify(|_, w| w.mode().i2c_master());
+            s.sercom.i2cm().ctrla.modify(|_, w| w.mode().i2c_master());
             // wait for configuration to take effect
-            while sercom.i2cm().syncbusy.read().enable().bit_is_set() {}
+            while !s.is_register_synced(i2c_reg_sync::Enable) {}
 
             // set the baud rate
             let gclk = clock.freq();
             let baud = (gclk.0 / (2 * freq.into().0) - 1) as u8;
-            sercom.i2cm().baud.modify(|_, w| w.baud().bits(baud));
+            s.sercom.i2cm().baud.modify(|_, w| w.baud().bits(baud));
 
-            sercom.i2cm().ctrla.modify(|_, w| w.enable().set_bit());
+            s.sercom.i2cm().ctrla.modify(|_, w| w.enable().set_bit());
             // wait for configuration to take effect
-            while sercom.i2cm().syncbusy.read().enable().bit_is_set() {}
+            while s.is_register_synced(i2c_reg_sync::Enable) {}
 
             // set the bus idle
-            sercom
+            s.sercom
                 .i2cm()
                 .status
                 .modify(|_, w| w.busstate().bits(BUS_STATE_IDLE));
             // wait for it to take effect
-            while sercom.i2cm().syncbusy.read().sysop().bit_is_set() {}
+            while !s.is_register_synced(i2c_reg_sync::SysOp) {}
         }
 
-        Self { sda, scl, sercom }
+        s
+    }
+
+    /// Wait for the registers synchronization
+    #[inline]
+    fn is_register_synced(&mut self, synced_reg: i2c_reg_sync) -> bool {
```

Comment:
not sure `&mut self` is needed for just a read

Reply:
Same suggestion as for the SPI and UART modules: consider using the approach that doesn't require maintaining a bespoke enum.